dhall-lsp-server/src/Dhall/LSP/Backend/Completion.hs

```diff
@@ -0,0 +1,186 @@
+module Dhall.LSP.Backend.Completion where
+
+import Data.Text (Text)
+import Dhall.LSP.Backend.Diagnostics (Position, positionToOffset)
+import System.Directory (doesDirectoryExist, listDirectory)
+import System.FilePath (takeDirectory, (</>))
+import System.Environment (getEnvironment)
+
+import Dhall.Context (empty, toList)
+import Data.List.NonEmpty (NonEmpty(..))
+import qualified Data.Text as Text
+import Dhall.Context (Context, insert)
+import Dhall.Core (Expr(..), Binding(..), Var(..), normalize, shift, Const(..), subst, pretty, reservedIdentifiers)
+import Dhall.TypeCheck (X(..), typeWithA, typeOf)
+import Dhall.Parser (Src, exprFromText)
+import qualified Dhall.Map
+import qualified Data.HashSet as HashSet
+
+import Dhall.LSP.Backend.Parsing (holeExpr)
+
+import Debug.Trace
+
+-- | Given the cursor position construct the corresponding 'completion query'
+-- consisting of the leadup, i.e. text leading up to the word prefix that is to
+-- be completed, as well as the prefix that is to be completed.
+completionQueryAt :: Text -> Position -> (Text, Text)
+completionQueryAt text pos = (completionLeadup, completionPrefix)
+ where
+  off = positionToOffset text pos
+  text' = Text.take off text
+  breakEnd :: (Char -> Bool) -> Text -> (Text, Text)
+  breakEnd p =
+    (\(l,r) -> (Text.reverse l, Text.reverse r)) . Text.break p . Text.reverse
+  (completionPrefix, completionLeadup) =
+    breakEnd (`elem` (" \t\n[(,=+*&|}#?>" :: [Char])) text'
+
+-- | A completion result, optionally annotated with type information.
+data Completion =
+  Completion {
+    completeText :: Text,
+    completeType :: Maybe (Expr Src X) }
+
+-- | Complete file names.
+completeLocalImport :: FilePath -> FilePath -> IO [Completion]
+completeLocalImport relativeTo prefix = do
+  let dir = takeDirectory relativeTo </> takeDirectory prefix
+  exists <- doesDirectoryExist dir
+  if not exists
+  then return []
+  else do
+    files <- listDirectory dir
```

Comment:
`listDirectory` is a potentially very expensive operation if the user auto-completes the wrong directory (think: `/nix/store/<tab>`), so I suggest timing out this operation if it takes too long so that the user's editor can recover gracefully.

Reply:
Hadn't thought of that, good point.